WSFrontend/org.openl.rules.ruleservice.ws/src/org/openl/rules/ruleservice/oauth/JWKInURLReader.java

```diff
@@ -0,0 +1,76 @@
+package org.openl.rules.ruleservice.oauth;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jwt.SignedJWT;
+
+/**
+ * Reader for jwk from url.
+ *
+ * @author ybiruk
+ */
+@Component
+public class JWKInURLReader implements JWKTokenReader {
+
+    private final Map<String, JWK> keyCache = new HashMap<>();
+
+    @Autowired
+    private Environment env;
+
+    /**
+     * Get jwk from url.
+     * Url can be passed in token header or set in properties.
+     *
+     * @return JWK
+     */
+    @Override
+    public JWK getJWK(SignedJWT jwt) {
+        try {
+            return getKey(jwt);
+        } catch (IOException | ParseException e) {
+            return null;
+        }
+    }
+
+    private JWK getKey(SignedJWT jwt) throws IOException, ParseException {
+        JWSHeader header = jwt.getHeader();
+        JWK jwk = keyCache.get(header.getKeyID());
+        if (jwk == null) {
+            // update cache loading jwk public key data from url
+            JWKSet jwkSet = loadJWKSet(header);
+            if (jwkSet != null) {
+                for (JWK key : jwkSet.getKeys()) {
+                    keyCache.put(key.getKeyID(), key);
+                }
+                jwk = keyCache.get(header.getKeyID());
+            }
+        }
+        return jwk;
+    }
+
+    private JWKSet loadJWKSet(JWSHeader header) throws IOException, ParseException {
+        URI jwkHeaderUrl = header.getJWKURL();
+        String jwkPropertyUrl = env.getProperty("ruleservice.authentication.jwks");
+        if (jwkHeaderUrl == null && jwkPropertyUrl == null) {
+            return null;
+        }
+        URL url = jwkHeaderUrl != null ? jwkHeaderUrl.toURL() : new URL(jwkPropertyUrl);
```

Comment:
jwkPropertyUrl can be empty or blank. Parse exception will be thrown in this case. Maybe it's better to check is it's blank in if condition above